src/u-root.md

```diff
@@ -46,416 +36,71 @@ attacks against shell interpreters[^11] and C programs[^8], we have started to
+Intel’s UEFI firmware. Buffer overflows in Intel’s UEFI and Active Management
+Technology (AMT) have also been discovered in several versions in recent years.
+
+Both UEFI[^12] and AMT[^4] are embedded operating systems, loaded from flash and
+possibly running network-facing software. Attacks against UEFI have extensively
+been studied[^9]. Most printers are network-attached and are a very popular
 exploitation target[^6]. Firmware is not visible to most users and is updated
-much less frequently (if at all) than programs. It is the first software to
+much less frequently (if at all) than OS software. It is the first software to
 run, at power on reset. Exploits in firmware are extremely difficult to detect,
 because firmware is designed to be as invisible as possible. Firmware is
 extremely complex; UEFI is roughly equivalent in size and capability to a Unix
 kernel. Firmware is usually closed and proprietary, with nowhere near the level
 of testing of kernels. These properties make firmware an ideal place for
-so-called advanced persistent threats[^10] [^18] [^5]. Once an exploit is
+so-called advanced persistent threats[^10] [^18] [^5] (APTs). Once an exploit is
 installed, it is almost impossible to remove, since the exploit can inhibit its
 removal by corrupting the firmware update process. The only sure way to mitigate
 a firmware exploit is to destroy the hardware.
 
-U-root is an excellent option for embedded systems. U-root contains only 5
-binaries, 4 of them from the Go toolchain, and the 5th is an init binary. The
-rest of the programs are contained in BIOS flash in source form, including
-packages. The search path is arranged so that when a command is invoked, if it
-is not in `/bin`, an installer is invoked instead which compiles the program
-into `/bin`. If the build succeeds, the command is executed. This first
-invocation takes a fraction of a second, depending on program complexity. After
-that, the RAM-based, statically linked binaries run in about a millisecond.
-Scripts are written in Go, not a shell scripting language, with two benefits:
-the shell can be simple, with fewer corner cases, and the scripting environment
-is substantially improved since Go is more powerful than most shell scripting
-languages, but also less fragile and less prone to parsing bugs.
-
-## U-root design
-
-The u-root boot image is a build toolchain and a set of programs in source
-form. When first used, a program and any needed but not-yet-built packages are
-built and installed, typically in a fraction of a second. With later uses, the
-binary is executed. The root file system is almost entirely unformed on boot;
-`/init` sets up the key directories and mounts, including common ones such as
-`/etc` and `/proc`.
-
-Since the init program itself is only 132 lines of code and is easy to change,
+## u-root design
+
+Designed to be a suitable option for embedded systems, the u-root build system
```

Comment:
Ah! Here it is a build system. It isn't possible to download official pre-built binaries, right?

Reply:
There is, though it's not being kept up to date. u-root is rather meant to be used for per-purpose customization, much like BusyBox and Buildroot.

I had set up https://github.com/linuxboot/u-root-builder with CI build artifacts attached, and later, Chris forked it (OOB, not on GitHub) to https://github.com/u-root/u-root-distro; it needs work to be really usable. This stuff was mainly used for workshops / trying things out.

Reply:
http://localhost:3000/u-root-qemu-demo.html this is based on the prebuilt binaries